File: Greedy/Medium/LET_no279_Perfect Squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        visited = [False] * (n + 1)
        squares = []
        for edge in range(1, int(n ** 0.5) + 1):
            square = edge * edge
            visited[square] = True
            squares.append(square)

        queue = collections.deque(squares)
        res = 1
        while queue:
            for _ in range(len(queue)):
                cur = queue.popleft()
                if cur == n:
                    return res
                for square in squares:
                    nxt = cur + square
                    if nxt <= n and not visited[nxt]:
                        visited[nxt] = True
                        queue.append(nxt)
            res += 1
        
        return 0

File: Greedy/Medium/LET_no279_Perfect Squares.cpp
class Solution {
public:
    int numSquares(int n) {
        vector<bool> visited(n + 1, false);
        vector<int> squares;
        queue<int> q;
        for (int i = 1; i * i <= n; i++) {
            int cur = i * i;
            squares.push_back(cur);
            visited[cur] = true;
            q.push(cur);
        }
        
        int res = 1;
        while (!q.empty()) {
            for (int i = q.size(); i > 0; i--) {
                int cur = q.front();
                q.pop();
                if (cur == n)
                    return res;
                
                for (int& num : squares) {
                    int nxt = cur + num;
                    if (nxt <= n && !visited[nxt]) {
                        visited[nxt] = true;
                        q.push(nxt);
                    }
                    else if (nxt > n)
                        break;
                }
            }
            res++;
        }
        return res;
    }
};